common/achievement/create.ts

```diff
@@ -0,0 +1,78 @@
+import { Prisma } from '@prisma/client';
+import { Achievement_template, JsonFilter } from '../../graphql/generated';
+import { ActionID, SpecificNotificationContext } from '../notification/actions';
+import { prisma } from '../prisma';
+import { TemplateSelectEnum, getAchievementTemplates } from './template';
+
+async function doesUserAchievementAlreadyExist<ID extends ActionID>(templateId: number, userId: string, context: SpecificNotificationContext<ID>) {
+    const keys = Object.keys(context);
+    const userAchievement = await prisma.user_achievement.findFirst({
+        where: {
+            templateId,
+            userId,
+            AND: keys.map((key) => {
+                return {
+                    context: {
+                        path: key,
+                        equals: context[key],
+                    },
+                };
+            }),
+        },
+        select: { id: true, userId: true, context: true, template: true, achievedAt: true, recordValue: true },
+    });
+    return userAchievement;
+}
+
+async function getOrCreateUserAchievement<ID extends ActionID>(template: Achievement_template, userId: string, context?: SpecificNotificationContext<ID>) {
+    const existingUserAchievement = await doesUserAchievementAlreadyExist(template.id, userId, context);
+    if (!existingUserAchievement) {
+        return await createAchievement(template, userId, context);
+    }
+    return existingUserAchievement;
+}
+
+async function createAchievement<ID extends ActionID>(templateToCreate: Achievement_template, userId: string, context: SpecificNotificationContext<ID>) {
+    const templatesByGroup = await getAchievementTemplates(TemplateSelectEnum.BY_GROUP);
+    const userAchievementsByGroup = await prisma.user_achievement.findMany({
+        where: { template: { group: templateToCreate.group } },
+        orderBy: { template: { groupOrder: 'asc' } },
+    });
+
+    const nextStepIndex = userAchievementsByGroup.length > 0 ? templateToCreate.groupOrder + 1 : 1;
```

Comment:
I don't feel comfortable in relying on the groupOrder field here. What if there is a gap?
We should find the array index of the last `userAchievements` within the `templatesByGroup` array, and then just pass the next one to `createNextUserAchievement`

Reply:
I just saw that `createNextUserAchievement` is recursive to get the same result.
I'm a little bit afraid what happens if the last achievement in the group is not active. This could crash the application 😅

Reply:
as mentioned in the comment before, we had this issue during implementation and changed it in a feature branch that is now merged. But your concerns are completely justified.